k8s/upgrade-job/src/helm/upgrade.rs

```diff
@@ -88,109 +81,108 @@ impl HelmUpgradeBuilder {
             .context(RegexCompile {
                 expression: umbrella_chart_regex.clone(),
             })?
             .is_match(chart.as_str())
         {
             chart_variant = HelmChart::Umbrella;
-            match self.umbrella_chart_dir {
-                Some(umbrella_dir) => chart_dir = umbrella_dir,
-                None => {
-                    return NoInputHelmChartDir {
-                        chart_name: UMBRELLA_CHART_NAME.to_string(),
-                    }
-                    .fail()
-                }
-            }
+
+            already_upgraded = to_version.eq(&from_version);
+            ensure!(already_upgraded, UmbrellaChartNotUpgraded);
         } else if Regex::new(core_chart_regex.as_str()) // Case: HelmChart::Core.
             .context(RegexCompile {
                 expression: core_chart_regex.clone(),
             })?
             .is_match(chart.as_str())
         {
             chart_variant = HelmChart::Core;
-            match self.core_chart_dir {
-                Some(core_dir) => chart_dir = core_dir,
-                None => {
-                    return NoInputHelmChartDir {
-                        chart_name: CORE_CHART_NAME.to_string(),
-                    }
-                    .fail()
-                }
-            }
-        } else {
-            // Case: Helm chart release is not a known helm chart installation.
-            return NotAKnownHelmChart { chart_name: chart }.fail();
-        }
-
-        let mut chart_yaml_path = chart_dir.clone();
-        chart_yaml_path.push("Chart.yaml");
-        let to_version = upgrade::path::version_from_chart_yaml_file(chart_yaml_path)?;
-        let from_version = upgrade::path::version_from_release_chart(chart)?;
-        let invalid_upgrades = upgrade::path::invalid_upgrade_path(self.upgrade_path_file)?;
 
-        // Check for already upgraded
-        let already_upgraded = to_version.eq(&from_version);
-        ensure!(!already_upgraded, InvalidUpgradePath);
+            // Skip upgrade-path validation and allow all upgrades for the Core helm chart, if the
+            // flag is set.
+            if !self.skip_upgrade_path_validation {
+                // Check for already upgraded
+                already_upgraded = to_version.eq(&from_version);
+
+                let upgrade_path_is_valid =
+                    upgrade::path::is_valid_for_core_chart(&from_version, self.upgrade_path_file)?;
+                ensure!(
+                    upgrade_path_is_valid || already_upgraded,
+                    InvalidUpgradePath
+                );
+            }
 
-        if !self.skip_upgrade_path_validation {
-            let upgrade_path_is_valid = upgrade::path::is_valid(
-                chart_variant.clone(),
+            // Generate args to pass to the `helm upgrade` command.
+            let mut values_yaml_path = chart_dir.clone();
+            values_yaml_path.push("values.yaml");
```

Comment:
Isn't set already set above, though we did None it?

Reply:
No... it isn't. Didn't get you.